### DISCUSSION TITLE
Use driver specific connections for guest-agent communication

### DISCUSSION THREAD

**Discussion by balajiv113** [lima-vm/lima#1925] [Ideas]
**Current Behaviour**
At present, Host <-> Guest agent communication is happening via ssh tunnel created for port forwarding. Basically, guest communication is another forwarded port for lima

**Proposal**
Decouple the port-forwarding and guest <-> host communication. For guest <-> host communication we will make use of native communication channels from each drivers.

- VZ supports vsock
- QEMU supports virtio-serial
Tested bi-directional communication with the below device
```
args = append(args, "-chardev", fmt.Sprintf("socket,path=%s,server=on,wait=off,id=qga0", vgaSock))
args = append(args, "-device", "virtio-serial")
args = append(args, "-device", "virtserialport,chardev=qga0,name=org.qemu.guest_agent.0")
```
- WSL2 supports vsock

With this approach, we would reduce the load on ssh master and also avoid the additional hop that's happening when passed via port-forwarding

**Comment by afbjorklund** [comment]
I would like the SSH communication to still exist, for other drivers. Similar to how we use sshfs, when no virtfs/virtiofs available ?

Specifically I was experimenting with external virtual Machines, using _only_ SSH. I suspect it would also break for VirtualBox...

**Reply by afbjorklund** [reply]
Fixed, now the other drivers still work like before:
* #2006

**Reply by afbjorklund** [reply]
Using virtio and vsock is still a good idea, where available.

Once upon a time, there was a similar discussion about 2376.
